tests/cpp/test_host_irs.cpp

```diff
@@ -387,6 +389,140 @@ INSTANTIATE_TEST_SUITE_P(
                                      : "useFusionExecutor";
     });
 
+// The following test simply demonstrate how to change current CUDA stream in
+// the host program
+TEST_F(NVFuserTest, HostIrSetStream) {
+  auto hic = std::make_unique<HostIrContainer>();
+  auto stream =
+      IrBuilder::create<StreamIr>(static_cast<IrContainer*>(hic.get()));
+  auto set_stream = IrBuilder::create<SetCurrentStream>(
+      static_cast<IrContainer*>(hic.get()), stream);
+  hic->pushBackTopLevelExprs(set_stream);
+
+  HostIrExecutor hie(std::move(hic));
+  setCurrentCUDAStream(c10::cuda::getDefaultCUDAStream(0));
+  hie.runWithInput({});
+  EXPECT_NE(
+      c10::cuda::getDefaultCUDAStream(0), c10::cuda::getCurrentCUDAStream(0));
+}
+
+using StreamHostIrTestParams = std::tuple<bool, int, int>;
+using StreamHostIrTest = NVFuserFixtureParamTest<StreamHostIrTestParams>;
+
+// The following test execute the same fusion `n_iterations` times by posting
+// the kernels on `n_streams` different streams in a Round-Robin fashion. We
+// thus produce `n_iterations` outputs from the same input, with a potential
+// overlap of n_streams/n_iterations
+TEST_P(StreamHostIrTest, SingleFusionMultipleStreams) {
+  auto [use_fusion_executor_cache, n_streams, n_iterations] = GetParam();
+
+  // [Step 1)] Define the Fusion we want to execute
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  std::vector<int64_t> input_sizes = {4, 8, 32};
+
+  auto tv0 = makeConcreteTensor(input_sizes);
+  auto tv1 = add(tv0, tv0);
+  auto tv2 = sum(tv1, {0});
+  fusion->addInput(tv0);
+  fusion->addOutput(tv2);
+
+  // [Step 2)] Instantiate an HostIroCntainer
+  auto hic = std::make_unique<HostIrContainer>();
+  FusionGuard::setCurFusion(hic.get());
+
+  // Create N different Streams
+  std::vector<StreamIr*> streams;
+  for (int i = 0; i < n_streams; i++) {
+    streams.push_back(
+        IrBuilder::create<StreamIr>(static_cast<IrContainer*>(hic.get())));
+  }
+
+  // [Step 3)] Create a HostUnit Ir holding the created fusion
+  auto host_unit = IrBuilder::create<HostUnit>(
+      static_cast<IrContainer*>(hic.get()), std::move(fusion));
```

Comment:
I'm surprised the static_cast is needed. Can't HostIrContainer be automatically upcasted to IrContainer to use https://github.com/NVIDIA/Fuser/blob/2310aec5c49aef3ee1806ac2b33b5d28dc27aedd/csrc/ir/builder.h#L44?

Reply:
I always had this issue, which I think comes from the compiler getting confused with upcasting and ellipsis

Here the type of compile error I get if I don't `static_cast` the ptr `hic.get()`
```
In file included from /opt/pytorch/Fuser/csrc/host_ir/host_ir.h:12,
                 from /opt/pytorch/Fuser/csrc/ir/all_nodes.h:10,
                 from /opt/pytorch/Fuser/csrc/device_lower/analysis/trivial_broadcast.h:11,
                 from /opt/pytorch/Fuser/csrc/compute_at_map.h:10,
                 from /opt/pytorch/Fuser/csrc/device_lower/lower2device.h:12,                                                                                   from /opt/pytorch/Fuser/csrc/evaluator_common.h:9,
                 from /opt/pytorch/Fuser/csrc/expr_evaluator.h:10,
                 from /opt/pytorch/Fuser/csrc/executor_kernel_arg.h:12,
                 from /opt/pytorch/Fuser/tests/cpp/test_host_irs.cpp:10:
/opt/pytorch/Fuser/csrc/ir/builder.h: In instantiation of ‘static T* nvfuser::IrBuilder::create(Args&& ...) [with T = nvfuser::hir::HostUnit; A
rgs = {nvfuser::hir::HostIrContainer*, std::unique_ptr<nvfuser::Fusion, std::default_delete<nvfuser::Fusion> >}]’:
/opt/pytorch/Fuser/tests/cpp/test_host_irs.cpp:105:47:   required from here
/opt/pytorch/Fuser/csrc/ir/builder.h:34:15: error: no matching function for call to ‘nvfuser::hir::HostUnit::HostUnit(nvfuser::IrBuilderPasskey
, nvfuser::hir::HostIrContainer*, std::unique_ptr<nvfuser::Fusion>)’
   34 |     T* node = new T(IrBuilderPasskey(fusion), std::forward<Args>(args)...);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

or
```
In file included from /opt/pytorch/Fuser/csrc/ir/all_nodes.h:10,
                 from /opt/pytorch/Fuser/csrc/device_lower/analysis/trivial_broadcast.h:11,
                 from /opt/pytorch/Fuser/csrc/compute_at_map.h:10,
                 from /opt/pytorch/Fuser/csrc/device_lower/lower2device.h:12,
                 from /opt/pytorch/Fuser/csrc/evaluator_common.h:9,
                 from /opt/pytorch/Fuser/csrc/expr_evaluator.h:10,
                 from /opt/pytorch/Fuser/csrc/executor_kernel_arg.h:12,
                 from /opt/pytorch/Fuser/tests/cpp/test_host_irs.cpp:10:
/opt/pytorch/Fuser/csrc/host_ir/host_ir.h:136:15: note:   inherited here
  136 |   using Expr::Expr;
      |               ^~~~
/opt/pytorch/Fuser/csrc/host_ir/host_ir.h:136:15: note:   candidate expects 4 arguments, 3 provided
/opt/pytorch/Fuser/csrc/host_ir/host_ir.h:137:3: note: candidate: ‘nvfuser::hir::SetCurrentStream::SetCurrentStream(nvfuser::IrBuilderPasskey, nvfuser::hir::StreamIr*)’
  137 |   SetCurrentStream(IrBuilderPasskey passkey, StreamIr* stream);
      |   ^~~~~~~~~~~~~~~~
/opt/pytorch/Fuser/csrc/host_ir/host_ir.h:137:3: note:   candidate expects 2 arguments, 3 provided
```

Reply:
I updated the tests with the new `createInContainer`